exporter/awsemfexporter/util.go

```diff
@@ -55,3 +60,72 @@ func replace(s, pattern string, value pdata.AttributeValue, logger *zap.Logger)
 	}
 	return strings.Replace(s, pattern, value.StringVal(), -1)
 }
+
+// getNamespace retrieves namespace for given set of metrics from user config.
+func getNamespace(rm *pdata.ResourceMetrics, namespace string) string {
+	if len(namespace) == 0 {
+		serviceName, svcNameOk := rm.Resource().Attributes().Get(conventions.AttributeServiceName)
+		serviceNamespace, svcNsOk := rm.Resource().Attributes().Get(conventions.AttributeServiceNamespace)
+		if svcNameOk && svcNsOk && serviceName.Type() == pdata.AttributeValueSTRING && serviceNamespace.Type() == pdata.AttributeValueSTRING {
+			namespace = fmt.Sprintf("%s/%s", serviceNamespace.StringVal(), serviceName.StringVal())
+		} else if svcNameOk && serviceName.Type() == pdata.AttributeValueSTRING {
+			namespace = serviceName.StringVal()
+		} else if svcNsOk && serviceNamespace.Type() == pdata.AttributeValueSTRING {
+			namespace = serviceNamespace.StringVal()
+		}
+	}
+
+	if len(namespace) == 0 {
+		namespace = defaultNamespace
+	}
+	return namespace
+}
+
+// getLogInfo retrieves the log group and log stream names from a given set of metrics.
+func getLogInfo(rm *pdata.ResourceMetrics, cWNamespace string, config *Config) (logGroup, logStream string) {
+	if cWNamespace != "" {
+		logGroup = fmt.Sprintf("/metrics/%s", cWNamespace)
+	}
+
+	// Override log group/stream if specified in config. However, in this case, customer won't have correlation experience
+	if len(config.LogGroupName) > 0 {
+		logGroup = replacePatterns(config.LogGroupName, rm.Resource().Attributes(), config.logger)
+	}
+	if len(config.LogStreamName) > 0 {
+		logStream = replacePatterns(config.LogStreamName, rm.Resource().Attributes(), config.logger)
+	}
+
+	return
+}
+
+// createMetricKey generates a hashed key from metric labels and additional parameters
```

Comment:
So it doesn't necessarily have to be in this PR but I do think we need to update this to use `label.Distinct` as we did in the statsdreceiver as its specially optimized for this use case

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1891#discussion_r551750484

Then we'll be able to use just a normal struct with the current `rateKeyParams` and a `label.Distinct`

Reply:
Thanks, I'll take a look the `label.Distinct`. We need to use `createMetricKey` for counter metric rate calculation in this PR as part of refactor.

Reply:
Hi @anuraaga, I tried to follow the idea that use a structure to define `rateKeyParams` fields along with `label.Distinct` that converted from `metric.LabelMap`, and then generate the metric rate calculation key string form both. The code changes had looked weird to me.  In my case, I have to convert the metric lablels `pdata.Metric.LabelMap` to `[]lable.KeyValue` so I can use `label.Distinct` sortting my labels. After that, I have to parse `label.Distinct`and `rateKeyParams` key/value pairs to calculate the rate key string like "paramkey:paramValue, labelkey:value,labelKey2:value2" which I think it is much complicated than the current `createMetricKey` implementation.  I feel this is not fitting my case unless I am using `label.Distinct` in the wrong way.  

I can convert the current `rateKeyParams` map to structure which makes sense to me, but not sure how I get the help with `label.Distinct` for my case.

Reply:
> After that, I have to parse label.Distinctand rateKeyParams key/value pairs to calculate the rate key string like "paramkey:paramValue, labelkey:value,labelKey2:value2"

I don't think you should need to compute a string anymore - once you have a struct with all the fields you need, including a `label.Distinct`, that struct could be the map key, not string, right?

Reply:
> that struct could be the map key, not string, right?

Reply:
Yeah - we already have it as a dep in the statsd receiver so I guess it's ok?